src/linear_algebra/clifford_algebra/basic.lean

```diff
@@ -64,23 +64,113 @@ end clifford_algebra
 
 /--
 The Clifford algebra of an `R`-module `M` equipped with a quadratic_form `Q`.
+
+We make it a structure to be sure that Lean can not unfold it, and we will also use
+irreducible operations for performance reasons.
 -/
-@[derive [inhabited, ring, algebra R]]
-def clifford_algebra := ring_quot (clifford_algebra.rel Q)
+structure clifford_algebra :=
+(out : ring_quot (clifford_algebra.rel Q))
 
 namespace clifford_algebra
 
+@[irreducible] private def zero : clifford_algebra Q := ⟨0⟩
+@[irreducible] private def one : clifford_algebra Q := ⟨1⟩
+@[irreducible] private def add : clifford_algebra Q → clifford_algebra Q → clifford_algebra Q
+| ⟨a⟩ ⟨b⟩ := ⟨a + b⟩
+@[irreducible] private def neg : clifford_algebra Q → clifford_algebra Q | ⟨a⟩ := ⟨-a⟩
+@[irreducible] private def mul : clifford_algebra Q → clifford_algebra Q → clifford_algebra Q
+| ⟨a⟩ ⟨b⟩ := ⟨a * b⟩
+@[irreducible] private def smul {S : Type*} [comm_semiring S] [algebra S (tensor_algebra R M)] :
+  S → clifford_algebra Q → clifford_algebra Q
+| r ⟨a⟩ := ⟨r • a⟩
```

Comment:
I think making this irreducible causes typeclass diamonds on things like the action of `units R` on `clifford_algebra Q`; this acts either directly via the units action, or via the units action on the tensor algebra.

We had this problem before for polynomial in #8392

Reply:
Currently it is irreducible (but after the fact), so it doesn't create problems (yet), but I agree that the example of polynomials gives a hint that it should probably be left reducible.

Reply:
I've changed this one (and only this one) to non-irreducible.